p2p/router.go

```diff
@@ -585,55 +585,82 @@ func (r *Router) openConnection(ctx context.Context, conn Connection) {
 func (r *Router) dialPeers() {
 	r.logger.Debug("starting dial routine")
 	ctx := r.stopCtx()
+
+	addresses := make(chan NodeAddress)
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < 4; i++ {
```

Comment:
I don't understand this? Why do we loop four times?

Reply:
just making 4 dialing threads.

Reply:
But why?

Reply:
reduce the likelihood that dialing and accepting peers will collide with each other.

Reply:
Tbh I don't know how I feel about this code...

Reply:
Starting a new goroutine for every attempted incoming *and* outgoing connections is good for one reason: it means that servers will be able to make progress towards being well connected in a network as fast as possible, and that other instances are potentially less likely to timeout connections while waiting for the other side to handshake or use the connection. This is potentially great.

It's also a major weakness for a couple of reason:
- there's a theoretical attack where seed nodes distribute addresses to peers that that will have a slow handshake while also dialing into those hosts, thereby provoking service to use resources to track both incoming and outgoing resources which will ultimately fail consuming 2x. Even if the nodes are fast, the the incoming and outgoing connections can block each other, and both can end up being dropped. The tests do this in normal operation. 
- even under normal situations, in large networks, starting up and attempting to dial many new hosts consumes a lot of resources, and could effectively lead to a "self-DoS"